slkspec/core.py

```diff
@@ -1,129 +1,355 @@
+    mode: str, default: rb
+        Specify the mode in which the files are opened
+
+        'r'       open for reading (default)
+        'b'       binary mode (default)
+        't'       text mode
+    file_permissions: int, default 0o3777
+        Permission when creating directories and files.
+    **kwargs:
+        Additional keyword arguments passed to the open file descriptor method.
+
+    Example
+    -------
+
+    Use fsspec to open data stored on tape, temporary data will be downloaded
+    to a central scratch folder:
+
+    ::
+
+        import ffspec
+        import xarray as xr
+
+        url = fsspec.open("slk:////arch/bb1203/data.nc",
+                          slk_cache="/scratch/b/b12346").open()
+        dset = xr.open_dataset(url)
+
+    """
+
+    write_msg: str = "Write mode is not suppored"
+    """Error message that is thrown if the files are attempted to be opened
+    in any kind of write mode."""
+
+    def __init__(
+        self,
+        url: str,
+        local_file: str,
+        *,
+        override: bool = True,
+        mode: str = "rb",
+        touch: bool = True,
+        file_permissions: int = 0o3777,
+        _lock: threading.Lock = _retrieval_lock,
+        _file_queue: Queue[Tuple[str, str]] = FileQueue,
+        **kwargs: Any,
+    ):
+        if not set(mode) & set("r"):  # The mode must have a r
+            raise NotImplementedError(self.write_msg)
+        if "b" not in mode:
+            kwargs.setdefault("encoding", "utf-8")
+        self._file = str(Path(local_file).expanduser().absolute())
+        self._url = str(url)
+        self.touch = touch
+        self.file_permissions = file_permissions
+        self._order_num = 0
+        self._file_obj: Optional[IO[Any]] = None
+        self._lock = _lock
+        self.kwargs = kwargs
+        self.mode = mode
+        self.newlines = None
+        self.error = "strict"
+        self.encoding = kwargs.get("encoding")
+        self.write_through = False
+        self._file_queue = _file_queue
+        print(self._file)
+        with _lock:
+            if not Path(self._file).exists() or override:
+                self._file_queue.put((self._url, str(Path(self._file).parent)))
+            elif Path(self._file).exists():
+                if self.touch:
+                    Path(self._file).touch()
+                self._file_obj = open(self._file, mode, **kwargs)
+
+    @property
+    def name(self) -> str:
+        """Get the file for the SLKFile object."""
+        if self._file_obj is not None:
+            return self._file
+        return self._url
+
+    def _retrieve_items(self, retrieve_files: list[tuple[str, str]]) -> None:
+        """Get items from the tape archive."""
+
+        retrieval_requests: Dict[Path, List[str]] = defaultdict(list)
+        logger.critical("Retrieving %i items from tape", len(retrieve_files))
+        for inp_file, out_dir in retrieve_files:
+            retrieval_requests[Path(out_dir)].append(inp_file)
```

Comment:
You might run a `parsers.slk_exists_bool(inp_file)` here for each file to check whether it really exists. It only works for explicite filenames (not regex or similar). This would prevent us from getting empty search id further below.

Reply:
I prefer to have an empty search, rather then checking each file individually and potentially slowing down the response.